tests/qnodes/test_qnode_reversible.py

```diff
@@ -0,0 +1,849 @@
+
+        assert positional_res == pytest.approx(list_res, abs=tol)
+        assert positional_grad == pytest.approx(list_grad, abs=tol)
+
+        array_res = circuit3(np.array([a, b, c]))
+        array_grad = circuit3.jacobian([np.array([a, b, c])])
+
+        list_res = circuit3([a, b, c])
+        list_grad = circuit3.jacobian([[a, b, c]])
+
+        assert positional_res == pytest.approx(array_res, abs=tol)
+        assert positional_grad == pytest.approx(array_grad, abs=tol)
+
+    @pytest.mark.parametrize("theta", thetas)
+    @pytest.mark.parametrize("G", [qml.ops.RX, qml.ops.RY, qml.ops.RZ])
+    def test_pauli_rotation_gradient(self, G, theta, tol):
+        """Tests that the automatic gradients of Pauli rotations are correct."""
+
+        def circuit(x):
+            qml.RX(x, wires=[0])
+            return qml.expval(qml.PauliZ(0))
+
+        dev = qml.device("default.qubit", wires=1)
+        circuit = ReversibleQNode(circuit, dev)
+
+        autograd_val = circuit.jacobian([theta])
+        manualgrad_val = (circuit(theta + np.pi / 2) - circuit(theta - np.pi / 2)) / 2
+        assert autograd_val == pytest.approx(manualgrad_val, abs=tol)
+
+    @pytest.mark.parametrize("theta", thetas)
+    def test_Rot_gradient(self, theta, tol):
+        """Tests that the automatic gradient of a arbitrary Euler-angle-parameterized gate is correct."""
+
+        def circuit(x, y, z):
+            qml.Rot(x, y, z, wires=[0])
+            return qml.expval(qml.PauliZ(0))
+
+        dev = qml.device("default.qubit", wires=1)
+        circuit = ReversibleQNode(circuit, dev)
+        eye = np.eye(3)
+
+        angle_inputs = np.array([theta, theta ** 3, np.sqrt(2) * theta])
+        autograd_val = circuit.jacobian(angle_inputs)
+        manualgrad_val = np.zeros((1, 3))
+
+        for idx in range(3):
+            onehot_idx = eye[idx]
+            param1 = angle_inputs + np.pi / 2 * onehot_idx
+            param2 = angle_inputs - np.pi / 2 * onehot_idx
+            manualgrad_val[0, idx] = (circuit(*param1) - circuit(*param2)) / 2
+
+        assert autograd_val == pytest.approx(manualgrad_val, abs=tol)
+
+    @pytest.mark.parametrize("op", [qml.CRX, qml.CRY, qml.CRZ,])
+    def test_controlled_rotation_gates_exception(self, op):
+        """Tests that an exception is raised when a controlled
+        rotation gate is used with the ReversibleQNode."""
+        # remove this test when this support is added
+        dev = qml.device("default.qubit", wires=2)
+
+        def circuit(x):
+            qml.PauliX(wires=0)
+            op(x, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        circuit = ReversibleQNode(circuit, dev)
+
+        with pytest.raises(
+            ValueError, match="Controlled-rotation gates are not currently supported"
+        ):
+            circuit.jacobian([0.542])
+
+    @pytest.mark.xfail
+    def test_controlled_RX_gradient(self, tol):
+        """Test gradient of controlled RX gate"""
```

Comment:
Wait, how is this possible? I thought controlled rotation gates should raise an error?

Reply:
```
Controlled-rotation gates are not currently supported with the reversible gradient method.
```

Reply:
Oh I just noticed the `xfail` 😆